### PR TITLE
Deployment refactoring and Docker build system

### PR DESCRIPTION
**Pull request by rafie**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -45,11 +45,12 @@ endif # SAN
 
 #----------------------------------------------------------------------------------------------
 
+ROOT=.
+
 ifeq ($(wildcard $(ROOT)/deps/readies/*),)
-$(info $(shell git submodule update --init --recursive &> /dev/null))
+___:=$(shell git submodule update --init --recursive &> /dev/null)
```
Comment (chayim) [inline Makefile:51]: What does __ do?

File: Makefile [hunk 2]
```diff
@@ -103,16 +104,24 @@ make cpp_tests     # run C++ tests (from tests/cpptests)
...
+
+common options for upload operations:
+  STAGING=1             # copy to staging lab area (for validation)
+  FORCE=1               # allow operation outside CI environment
+  VERBOSE=1             # show more details
+  NOP=1                 # do not copy, just print commands
 
 make docs          # create documentation
-make deploydocs    # deploy documentation
+make deploy-docs   # deploy documentation
```
Comment (chayim) [inline Makefile:123]: Yay :D

File: build/docker/dockerfile.tmpl [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+RUN make build SHOW=1
+
+{% if PACK == "1" %}
+RUN make build LITE=1 SHOW=1
+RUN make build COORD=rlec SHOW=1
+{% endif %}
+
+{% if TEST == "1" %}
+RUN make build COORD=oss SHOW=1
```
Comment (chayim) [inline build/docker/dockerfile.tmpl:30]: It's awkware that TEST specifies the COORD. Maybe that should be separate? WDYT?

File: build/docker/dockerfile.tmpl [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+RUN ./deps/readies/bin/getpy2
+RUN ./sbin/system-setup.py
+
+RUN /usr/local/bin/redis-server --version
+RUN make fetch
+
+RUN make build SHOW=1
+
+{% if PACK == "1" %}
+RUN make build LITE=1 SHOW=1
```
Comment (chayim) [inline build/docker/dockerfile.tmpl:25]: Can these two be combined to make build LITE=1 SHOW=1 COORD=rlec?

File: build/docker/dockerfile.tmpl [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+RUN mkdir -p bin/artifacts
+RUN make pack
+RUN make pack COORD=rlec
+RUN make pack LITE=1
+RUN set -e ;\
+    echo "Artifacts:" ;\
+    du -ah --apparent-size /build/bin/artifacts
+{% endif %}
+
+{% if TEST == "1" %}
```
Comment (chayim) [inline build/docker/dockerfile.tmpl:43]: Maybe put all the TEST together, and all the pack. WDYT?

File: sbin/upload-artifacts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+[[ $VERBOSE == 1 ]] && du -ah --apparent-size *
+
+#----------------------------------------------------------------------------------------------
+
+s3_upload_file() {
+	local file="$1"
+	local s3_dir="$2"
+	[[ $s3_dir != */ ]] && s3_dir="${s3_dir}/"
+	
+	$OP python2 -m awscli s3 cp $file $s3_dir --acl public-read --no-progress
```
Comment (chayim) [inline sbin/upload-artifacts:90]: You probably can call awscli directly without python2 here.

**Review by chayim — CHANGES_REQUESTED** [review]
Small changes - mostly a curiosity.

File: build/docker/dockerfile.tmpl (hunk 4 above)
Comment (rafie) [reply build/docker/dockerfile.tmpl:25]: Not likely. Those are two variations of the same build. I can maybe offer `make ALL=1`?

File: build/docker/dockerfile.tmpl (hunk 3 above)
Comment (rafie) [reply build/docker/dockerfile.tmpl:30]: The OSS coordinator is sort of missing in action in the Docker world (there is no OSS coordinator in Docker at the moment). However, if we request the Docker build to include tests, we run the full set of tests, which includes the OSS coordinator.

File: sbin/upload-artifacts (hunk 6 above)
Comment (rafie) [reply sbin/upload-artifacts:90]: Yes, I know. I think I reacted to a plain `awscli`. I'll revert this after this is more stable.

File: build/docker/dockerfile.tmpl (hunk 4 above)
Comment (chayim) [reply build/docker/dockerfile.tmpl:25]: Agreed

**Review by chayim — DISMISSED** [review]

**Review by oshadmi — APPROVED** [review]
👏🏼 👍🏼

**Pull request merged**